Use Maven Wrapper to run TCK FATs

Azquelt opened the pull request:
Fixes #24289

Builds on previous work in #25005.

- Add the maven wrapper script and properties file to the fattest.simplicity project
- When running a TCK, copy these files to the test directory and use them to launch the TCK test
- When running in CI, download the maven distribution from an internal repo instead
- Refactor some methods out of TCK Runner
- Improve the timeout implementation when running a TCK (computation of the timeout value may still need some work)

tevans78 committed:
Use mvn wrapper, downloading binaries from artifactory

This reverts commit 3794c6106212d5c44864ec6ff41b9a953e29e1d6.

abutch3r committed:
Improve downloadFile resilience

fix for graphQL MVN goal fat_tck

Azquelt committed:
WIP: Fixes to use mvn wrapper

Use the standard .m2 directory if the user isn't configured to use
artifactory. Use standard mirror configuration in the settings.xml in
this case.

Keep the default values in maven-wrapper.properties, only overwrite them
if the user has configured artifactory.

Don't use Machine.execute because it merges arguments and then splits
them on whitespace. This requires our own handling of the process
output which we do asynchronously so that we can kill off the process if
it takes too long.

Split the maven output into lines as we read it to avoid repeatedly
splitting and merging it.

Azquelt committed:
Undo change to GraphQL repeat ID

This change shouldn't be necessary now that we pass arguments to maven
sensibly and undoing it serves as a good test.

Azquelt committed:
Fix zOS ascii conversion

Azquelt committed:
Cleanup, javadoc and parameter checks

Azquelt committed:
Revert simplicity run process timeout changes

We're no longer using the simplicity code for other reasons, so remove
the changes to respect timeout. The changes may still be good but
they'll be done and tested in another PR.

Azquelt committed:
Add conditional settings argument

Azquelt committed:
More fixes to the TCK wrapper support

- Always overwrite files when exporting from classpath (ensures we
  refresh settings.xml so we can assume it always needs updated)
- Add hash checking when downloading files, along with better error
  logging.

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+                    Log.info(c, "zosTagASCII", line);
+                }
+
+                throw new Exception("Process failure <chtag> see log for output.");
```
Maybe include output in exception message

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+                throw new Exception("Process failure <chtag> see log for output.");
+            }
+        } catch (Exception e) {
+            Log.error(c, "Could not tag zos file as ASCII", e);
```
Re-throw the exception ... no way it will recover

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+    public static Properties updateWrapperPropertiesFile(File wrapperPropertiesFile) throws IOException {
+        Properties props = new Properties();
+        //get the artifactory server
+        String artifactoryServer = System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_SERVER_KEY);
```
Pass in server and update doc to be clear this is only run when using artifactory

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+    public static boolean useArtifactory() {
+        Log.info(c, "useArtifactory", "Force external: " + System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_FORCE_EXTERNAL_KEY, ""));
+        boolean forceExternal = Boolean.getBoolean(FAT_TEST_PREFIX + ARTIFACTORY_FORCE_EXTERNAL_KEY);
+        if (forceExternal) {
+            return false;
+        }
+
+        return !System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_SERVER_KEY, "").isEmpty();
```
rewrite to make more readable ... single return?

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+        return !System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_SERVER_KEY, "").isEmpty();
+    }
+
+    public static String getArtifactoryServer() {
+        return System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_SERVER_KEY);
+    }
+
+    public static String getArtifactoryUser() {
+        return System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_USER_KEY);
+    }
+
+    public static String getArtifactoryToken() {
+        return System.getProperty(FAT_TEST_PREFIX + ARTIFACTORY_TOKEN_KEY);
+    }
```
Use these methods where appropriate

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+     */
+    public static File getTemporaryMavenHomeDir() throws IOException {
+        // user.dir is set to the test project directory when running a FAT
+        String userDir = System.getProperty("user.dir");
```
consider using a different starting point

Azquelt replied:
On re-review I think this is actually correct, `user.dir` is the current working directory.

However, `File` will automatically resolve against this directory when an absolute file path is needed so I'm removing the manual system property lookup and adding some validation of the final directory.

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKUtilities.java:
```diff
@@ -207,4 +220,542 @@ public static TCKJarInfo parseTCKDependencies(Type type, String[] dependencyOutp
+     * @param  directory   the file to check
+     * @throws IOException if the file does not exist or is not a directory
+     */
+    public static void requireDirectory(File directory) throws IOException {
+        requireExists(directory);
+        if (!directory.isDirectory()) {
+            throw new IOException(directory + " is not a directory");
+        }
+    }
```
Use this method where appropriate

tevans78 reviewed (dismissed):
No major changes required ... small issues can be done as a follow up.

Azquelt committed:
Review changes for TCK wrapper support

Azquelt committed:
Reduce GraphQL and Rest Client TCK heap size

- GraphQL server heap 2G -> 1.5G
- Rest Client TCK test process heap 2G -> 1.5G

See if this fixes failures when starting the JVM on z/OS because a 2G
heap can't be allocated.

Azquelt commented:
#libby
#build

tevans78 commented on dev/fattest.simplicity/src/componenttest/topology/utils/tck/TCKRunner.java:
```diff
@@ -880,71 +911,45 @@ private static String jarPathInDir(String jarNameFragment, String dir) {
-                lastLines.add(output.get(lineNum));
-                lineNum++;
+        if (mavenUserHome != null) {
+            String artifactoryServer = TCKUtilities.getArtifactoryServer();
```
I think we looked at this before but should there be a `TCKUtilities.useArtifactory()` guard around these blocks?

tevans78 replied:
`mavenUserHome` will always be null if `useArtifactory()` returned false ... so the code is correct, just perhaps needs a comment

Azquelt commented:
Failures in the build are not related to this issue.
